pypesto/sampling/pymc3.py

```diff
@@ -135,21 +135,21 @@ class TheanoLogProbability(tt.Op):
 
     def __init__(self, problem: Problem, beta: float = 1.):
         self._objective: Objective = problem.objective
-
-        # initialize the log probability Op
-        self._log_prob = \
-            lambda x: - beta * self._objective(x, sensi_orders=(0,))
-
+        self._beta: float = beta
+        self._cached_theta = None
+        self._cached_dlogp = None
         # initialize the sensitivity Op
         if problem.objective.has_grad:
-            self._log_prob_grad = TheanoLogProbabilityGradient(problem, beta)
+            self._log_prob_grad = TheanoLogProbabilityGradient(self)
         else:
             self._log_prob_grad = None
 
     def perform(self, node, inputs, outputs, params=None):
         theta, = inputs
-        log_prob = self._log_prob(theta)
-        outputs[0][0] = np.array(log_prob)
+        logp, dlogp = self._objective(theta, sensi_orders=(0, 1))
```

Comment:
I think it would make more sense to cache values from the gradient -> prob, not prob -> grad. this is because whenever we evaluate the gradient , amici also computes the function value, but not the other way round. thus, when using a gradient free sampler like `pypesto.Pymc3sample(step_function=pymc3.Slice)`, now always the gradient is calculated, which is however never required, if I see correctly.

Reply:
Indeed with this change the gradient is computed also for samplers which do not need it, so it is not ideal. But it looks like that the gradient is always computed after the function value, so if we cached inside the gradient there would be no cache hits I am afraid. Then, probably having two versions is the best way to do it.

Reply:
Agreeing on that. It might be difficult to find out whether gradients are required however. If nothing is specified, we try to use the Nuts sampler, with fallback to Slice if grad=NullType, i.e. when not objective.has_grad(). So I guess in that case one can just preemptively check objective.has_grad() and generate the log_prob accordingly. If a sampler is explicitly specified, this will not be so easy, one would need to figure out which samplers in pymc3 use gradients and which not ...